### PR TITLE
Storage from_file method

### LINKED ISSUE #884: Support for memory-mapped files
In Lua torch tensor constructors accepted filenames, and could use memory mapped files as their data.

[Forum thread](https://discuss.pytorch.org/t/quickly-loading-raw-binary-files/797)

### PR DESCRIPTION
**Pull request by vlasenkov**
Resolves #884

### CODE REVIEW

File: torch/csrc/generic/StorageMethods.cpp [hunk 1]
```diff
@@ -179,6 +179,31 @@ static PyObject * THPStorage_(fromBuffer)(PyObject *_unused, PyObject *args, PyO
...
+  const char *filename;
+  Py_ssize_t size = 0;
+  int shared = 0, shared_mem = 0;
+  static char *kwlist[] = {"filename", "shared", "size", "shared_mem", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, keywds, "s|pnp", kwlist,
+              &filename, &shared, &size, &shared_mem)) {
+    return NULL;
+  }
+#if defined(__APPLE__)
+  shared_mem = 0;
```
Comment (apaszke) [inline torch/csrc/generic/StorageMethods.cpp]: Actually why do we have to set this to 0 on macOS?

File: torch/csrc/generic/StorageMethods.cpp [hunk 2]
```diff
@@ -179,6 +179,31 @@ static PyObject * THPStorage_(fromBuffer)(PyObject *_unused, PyObject *args, PyO
 }
 #endif
 
+static PyObject * THPStorage_(fromFile)(PyObject *_unused, PyObject *args, PyObject *keywds)
+{
+  HANDLE_TH_ERRORS
+  const char *filename;
+  Py_ssize_t size = 0;
+  int shared = 0, shared_mem = 0;
+  static char *kwlist[] = {"filename", "shared", "size", "shared_mem", NULL};
```
Comment (apaszke) [inline torch/csrc/generic/StorageMethods.cpp]: can you change `shared_mem` to `shm`? It's unclear how `shared` and `shared_mem` is different (and the difference is that it uses the shm subsystem)

File: torch/csrc/generic/StorageMethods.cpp [hunk 3]
```diff
@@ -179,6 +179,31 @@ static PyObject * THPStorage_(fromBuffer)(PyObject *_unused, PyObject *args, PyO
...
+  if (!PyArg_ParseTupleAndKeywords(args, keywds, "s|pnp", kwlist,
```
Comment (apaszke) [inline torch/csrc/generic/StorageMethods.cpp]: `p` format is supported only in Python 3.3+

File: test/test_torch.py [hunk 4]
```diff
@@ -3305,6 +3306,44 @@ def test_from_buffer(self):
...
 
+    def test_from_file(self):
+        size = 10000
+        filename = 'testPytorchStorageFromFile'
+        sys = platform.system()
+
+        def check(filename, share_mem):
+            s1 = torch.FloatStorage.from_file(filename, True, size, share_mem)
+            t1 = torch.FloatTensor(s1).copy_(torch.randn(size))
+            t1_copy = t1.new().resize_as_(t1).copy_(t1)
```
Comment (apaszke) [inline test/test_torch.py]: this is the same as `t1.clone()`

File: test/test_torch.py [hunk 5]
```diff
@@ -3305,6 +3306,44 @@ def test_from_buffer(self):
...
+            if share_mem and sys == 'Linux':
+                filename = os.path.join('/dev/shm', filename)
+
+            # check mapping
+            s2 = torch.FloatStorage.from_file(filename, True, size)
+            t2 = torch.FloatTensor(s2)
+            t2_copy = t2.new().resize_as_(t2).copy_(t2)
+            self.assertEqual(t1_copy, t2_copy, 0)
```
Comment (apaszke) [inline test/test_torch.py]: Why do you need these clones? Can't you just compare `t1` and `t2` directly?

File: test/test_torch.py [hunk 6]
```diff
@@ -3305,6 +3306,44 @@ def test_from_buffer(self):
...
+            t1.fill_(rnum)
+            self.assertEqual(t1, t2, 0)
+
+            # check changes to t2 from t1
+            rnum = random.uniform(-1, 1)
+            t2.fill_(rnum)
+            self.assertEqual(t1, t2, 0)
+
+            if sys.startswith('win'):
+                os.remove(filename)
```
Comment (apaszke) [inline test/test_torch.py]: why is this necessary? did you test it on windows?

File: test/test_torch.py [hunk 7]
```diff
@@ -3305,6 +3306,44 @@ def test_from_buffer(self):
         self.assertEqual(floats.size(), 1)
         self.assertEqual(floats[0], 2.25)
 
+    def test_from_file(self):
+        size = 10000
+        filename = 'testPytorchStorageFromFile'
+        sys = platform.system()
```
Comment (apaszke) [inline test/test_torch.py]: it's a bad variable name, because it shadows the `sys` module (btw you can just use `sys.platform` for this)

File: test/test_torch.py [hunk 8]
```diff
@@ -3305,6 +3306,44 @@ def test_from_buffer(self):
...
+
+            if sys.startswith('win'):
+                os.remove(filename)
+
+        # check open + open
+        check(filename, False)
+
+        # check shm_open + open
+        if sys != 'Darwin':  # OS X has no /dev/shm
+            check(filename, True)
```
Comment (apaszke) [inline test/test_torch.py]: this should be wrapped in a `try ... except ... finally ...` where finally ensure that the file is removed at the end of the test.

File: test/test_torch.py [hunk 9]
```diff
@@ -3305,6 +3306,44 @@ def test_from_buffer(self):
...
+        size = 10000
+        filename = 'testPytorchStorageFromFile'
+        sys = platform.system()
+
+        def check(filename, share_mem):
+            s1 = torch.FloatStorage.from_file(filename, True, size, share_mem)
+            t1 = torch.FloatTensor(s1).copy_(torch.randn(size))
+            t1_copy = t1.new().resize_as_(t1).copy_(t1)
+            if share_mem and sys == 'Linux':
+                filename = os.path.join('/dev/shm', filename)
```
Comment (apaszke) [inline test/test_torch.py]: Can't you just pass `share_mem` to `t2` as well? I wouldn't like to depend on these files being in `/dev/shm`

Comment (vlasenkov) [inline test/test_torch.py]: this will lead to what we discussed already in the issue thread. to creation of two shared memory storages

File: torch/csrc/generic/StorageMethods.cpp (hunk 1 above)
Comment (vlasenkov) [inline torch/csrc/generic/StorageMethods.cpp]: To completely prohibit using this argument on Mac. Because it is useless on mac

File: torch/csrc/generic/StorageMethods.cpp (hunk 2 above)
Comment (vlasenkov) [inline torch/csrc/generic/StorageMethods.cpp]: I chose this naming to be the same as in Lua Torch. Anyway we should write description in docs. But if you think `shm` is better I'll change

Comment (vlasenkov) [inline torch/csrc/generic/StorageMethods.cpp]: By the way, where should be Storage docs stored? In the `_torch_docs.py`, or some new `_storage_docs.py`?

File: torch/csrc/generic/StorageMethods.cpp (hunk 1 above)
Comment (apaszke) [inline torch/csrc/generic/StorageMethods.cpp]: Why is it useless? Isn't the filed unlinked when the storage is destructed? In theory you could open multiple views

File: torch/csrc/generic/StorageMethods.cpp (hunk 2 above)
Comment (apaszke) [inline torch/csrc/generic/StorageMethods.cpp]: I think it's better to use `shm`

File: test/test_torch.py (hunk 9 above)
Comment (apaszke) [inline test/test_torch.py]: Uh because `shm=True` unlinks the file, while `shm=False` doesn't, right?

**Comment by apaszke** [comment]
Actually now I think it should use `libshm` even now. Otherwise the unlinking semantics are going to change, and later commits will be breaking.

File: test/test_torch.py (hunk 9 above)
Comment (vlasenkov) [inline test/test_torch.py]: yes

**Comment by vlasenkov** [comment]
So, I'm removing `shared_mem` for now to save at least ordinary `mmap` functionality

**Comment by apaszke** [comment]
Sounds good

**Comment by vlasenkov** [comment]
where should be Storage docs stored? In the _torch_docs.py, or some new _storage_docs.py?

**Comment by apaszke** [comment]
I think you need to create `_storage_docs.py`. We only document functions implemented in Python at the moment

**Comment by vlasenkov** [comment]
@apaszke, I've added docs. ready for review/merge

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thank you!
